src/query/api/v1/middleware/query_inspection.go

```diff
@@ -0,0 +1,182 @@
+	rangeThresholdMetricName = "range_threshold"
+	querySizeSmall           = "small"
+	querySizeLarge           = "large"
+
+	// promDefaultLookback is the default lookback in Prometheus, that affects
+	// the actual range it fetches datapoints from.
+	promDefaultLookback = time.Minute * 5
+)
+
+type queryInspectionMetrics struct {
+	notQuery            tally.Counter
+	notInspected        tally.Counter
+	belowCountThreshold tally.Counter
+	badQuery            tally.Counter
+	belowRangeThreshold tally.Counter
+	largeTags           tally.Counter
+}
+
+func newQueryInspectionMetrics(scope tally.Scope) queryInspectionMetrics {
+	buildCounter := func(status string) tally.Counter {
+		return scope.Tagged(map[string]string{"status": status}).Counter("count")
+	}
+
+	return queryInspectionMetrics{
+		notQuery:            buildCounter("notQuery"),
+		notInspected:        buildCounter("notInspected"),
+		belowCountThreshold: buildCounter("belowCountThreshold"),
+		badQuery:            buildCounter("badQuery"),
+		belowRangeThreshold: buildCounter("belowRangeThreshold"),
+		largeTags:           buildCounter("largeTags"),
+	}
+}
+
+func retrieveQueryRange(expr parser.Node, rangeSoFar time.Duration) time.Duration {
+	var (
+		queryRange  = rangeSoFar
+		useLookback = rangeSoFar == 0
+	)
+
+	parser.Inspect(expr, func(node parser.Node, path []parser.Node) error {
+		switch n := node.(type) {
+		case *parser.SubqueryExpr:
+			useLookback = false
+			d := retrieveQueryRange(n.Expr, n.Range+n.Offset+rangeSoFar)
+			if d > queryRange {
+				queryRange = d
+			}
+		case *parser.MatrixSelector:
+			useLookback = false
+			// nolint
+			v := n.VectorSelector.(*parser.VectorSelector)
+			if d := v.Offset + n.Range + rangeSoFar; d > queryRange {
+				queryRange = d
+			}
+		case *parser.VectorSelector:
+			if d := n.Offset + rangeSoFar; d > queryRange {
+				queryRange = d
+			}
+		}
+
+		return nil
+	})
+
+	if useLookback {
+		// NB: no range provided in any query selecturs; use lookback time.
+		queryRange += promDefaultLookback
+	}
+
+	return queryRange
+}
+
+func isQueryPath(path string) bool {
+	idx := strings.LastIndex(path, "/")
+	if idx < 0 {
+		return false
+	}
+
+	path = path[idx+1:]
+	return path == "query" || path == "query_range"
+}
+
+// NB: inspectQuerySize determines this query's size; if the number of fetched
+// series exceeds a certain number, and the query range exceeds a certain range,
+// the query is classified as "large"; otherwise, this query is "small".
+func inspectQuerySize(
+	w http.ResponseWriter,
+	r *http.Request,
+	path string,
+	metrics queryInspectionMetrics,
+	cfg *config.MiddlewareConfiguration,
+) map[string]string {
+	tags := map[string]string{
+		querySizeMetricName:      querySizeSmall,
+		countThresholdMetricName: "0",
+		rangeThresholdMetricName: "0",
+	}
+
+	// NB: query categorization is only relevant for query and query_range endpoints.
+	if !isQueryPath(path) {
+		metrics.notQuery.Inc(1)
+		return tags
+	}
```

Comment:
If it's not a query path, can we just return nil or an empty map? Would be ideal to not have any extra tags on those.